google-cloud-firestore/lib/google/cloud/firestore/batch.rb

```diff
@@ -0,0 +1,1020 @@
+# Copyright 2017, Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/firestore/document"
+require "google/cloud/firestore/collection"
+require "google/cloud/firestore/query"
+require "google/cloud/firestore/convert"
+
+module Google
+  module Cloud
+    module Firestore
+      ##
+      # # Batch
+      #
+      # A batch in Cloud Firestore is a set of writes that execute
+      # atomically at a single logical point in time in a database.
+      #
+      # All changes are accumulated in memory until the block passed to
+      # {Database#batch} completes. Unlike transactions, batches are not
+      # automatically retried. See {Database#batch}.
+      #
+      # Batched writes have fewer failure cases than transactions and use
+      # simpler code. They are not affected by contention issues, because they
+      # don't depend on consistently reading any documents.
+      #
+      # @see https://firebase.google.com/docs/firestore/manage-data/transactions
+      #   Transactions and Batched Writes
+      #
+      # @example
+      #   require "google/cloud/firestore"
+      #
+      #   firestore = Google::Cloud::Firestore.new
+      #
+      #   firestore.batch do |b|
+      #     # Set the data for NYC
+      #     b.col("cities").doc("NYC").set({ name: "New York City" })
+      #
+      #     # Update the population for SF
+      #     b.col("cities").doc("SF").update({ population: 1000000 })
+      #
+      #     # Delete LA
+      #     b.col("cities").doc("LA").delete
+      #   end
+      #
+      class Batch
+        ##
+        # New Batch object.
+        def initialize
+          @writes = []
+        end
+
+        ##
+        # The project resource the Cloud Firestore batch belongs to.
+        #
+        # @return [Project] project resource
+        def project
+          @database.project
+        end
+
+        ##
+        # The database resource the Cloud Firestore batch belongs to.
+        #
+        # @return [Database] database resource.
+        def database
```

Comment:
This is a bit of future proofing, as my understand is that users will be able to create multiple databases in the future, and we would benefit by being more explicit. But we can alias this method as `firestore` if you would prefer that.

Reply:
Do we have a precedent for this? I think it might make more sense to have to distinct Firestore clients when we have multiple databases. This is what we do for the Firebase RTDB SDK for example.